# Introduction to Python for ML Quiz - 2021 Hamoye Intern

In [1]:
#importing libraries
import pandas as pd
import numpy as np

In [2]:
# Data set import
url = 'https://raw.githubusercontent.com/WalePhenomenon/climate_change/master/fuel_ferc1.csv'
fuel_data = pd.read_csv(url, error_bad_lines=False)
fuel_data.describe(include='all')

record_id  utility_id_ferc1   report_year  \
count                     29523      29523.000000  29523.000000   
unique                    29523               NaN           NaN   
top     f1_fuel_2006_12_186_3_2               NaN           NaN   
freq                          1               NaN           NaN   
mean                        NaN        118.601836   2005.806050   
std                         NaN         74.178353      7.025483   
min                         NaN          1.000000   1994.000000   
25%                         NaN         55.000000   2000.000000   
50%                         NaN        122.000000   2006.000000   
75%                         NaN        176.000000   2012.000000   
max                         NaN        514.000000   2018.000000   

       plant_name_ferc1 fuel_type_code_pudl fuel_unit  fuel_qty_burned  \
count             29523               29523     29343     2.952300e+04   
unique             2315                   6         9              NaN   
top           big stone                 gas       mcf              NaN   
freq                156               11486     11354              NaN   
mean                NaN                 NaN       NaN     2.622119e+06   
std                 NaN                 NaN       NaN     9.118004e+06   
min                 NaN                 NaN       NaN     1.000000e+00   
25%                 NaN                 NaN       NaN     1.381700e+04   
50%                 NaN                 NaN       NaN     2.533220e+05   
75%                 NaN                 NaN       NaN     1.424034e+06   
max                 NaN                 NaN       NaN     5.558942e+08   

        fuel_mmbtu_per_unit  fuel_cost_per_unit_burned  \
count          29523.000000               29523.000000   
unique                  NaN                        NaN   
top                     NaN                        NaN   
freq                    NaN                        NaN   
mean               8.492111                 208.649031   
std               10.600220                2854.490090   
min                0.000001                -276.080000   
25%                1.024000                   5.207000   
50%                5.762694                  26.000000   
75%               17.006000                  47.113000   
max              341.260000              139358.000000   

        fuel_cost_per_unit_delivered  fuel_cost_per_mmbtu  
count                   2.952300e+04         29523.000000  
unique                           NaN                  NaN  
top                              NaN                  NaN  
freq                             NaN                  NaN  
mean                    9.175704e+02            19.304354  
std                     6.877593e+04          2091.540939  
min                    -8.749370e+02           -41.501000  
25%                     3.778500e+00             1.940000  
50%                     1.737100e+01             4.127000  
75%                     4.213700e+01             7.745000  
max                     7.964521e+06        359278.000000

In [3]:
#This code gives details of the datatype of each column
fuel_data.dtypes

record_id                        object
utility_id_ferc1                  int64
report_year                       int64
plant_name_ferc1                 object
fuel_type_code_pudl              object
fuel_unit                        object
fuel_qty_burned                 float64
fuel_mmbtu_per_unit             float64
fuel_cost_per_unit_burned       float64
fuel_cost_per_unit_delivered    float64
fuel_cost_per_mmbtu             float64
dtype: object

### Question 3: Which of the following fuel type code has the lowest average fuel cost per unit burned?

In [7]:
fuel_data.groupby('fuel_type_code_pudl')['fuel_cost_per_unit_burned'].min()

fuel_type_code_pudl
coal         0.00
gas          0.00
nuclear      0.00
oil          0.00
other     -119.04
waste     -276.08
Name: fuel_cost_per_unit_burned, dtype: float64

#### From the analysis above the answer is WASTE

### Question 4: What is the standard deviation and 75th percentile of the measure of energy per unit (Fuel_mmbtu_per_unit) in two decimal places?

In [8]:
fuel_data['fuel_mmbtu_per_unit'].std()

10.600220307806886

In [9]:
np.percentile(fuel_data['fuel_mmbtu_per_unit'], 75)

17.006

#### Answer: 10.60 and 17.01

### Question 5: What is the skewness and kurtosis for the fuel quantity burned in two decimal places?

In [10]:
fuel_data['fuel_qty_burned'].skew()

15.851495469109503

In [11]:
fuel_data['fuel_qty_burned'].kurtosis()

651.3694501337732

#### Answer: 15.85 and 651.37

### Question 6: Which feature has missing values and what is the total number of missing value and percentage of the missing rows as a factor of the total number of rows in three decimal places?

In [12]:
fuel_data.isnull().sum()

record_id                         0
utility_id_ferc1                  0
report_year                       0
plant_name_ferc1                  0
fuel_type_code_pudl               0
fuel_unit                       180
fuel_qty_burned                   0
fuel_mmbtu_per_unit               0
fuel_cost_per_unit_burned         0
fuel_cost_per_unit_delivered      0
fuel_cost_per_mmbtu               0
dtype: int64

In [17]:
perc_of_missing = (180/29523)*100
print(perc_of_missing)

0.609694136774718


#### Feature: Fuel Unit, Total: 180, Percent: 0.610

### Question 7: The feature with missing values falls under what category? What missing value imputation technique would you use?

### Question 8: Which of the features has the second and third lowest correlation with the Fuel Cost Per Unit Burned?

In [30]:
np.correlate(fuel_data['utility_id_ferc1'], fuel_data['fuel_cost_per_unit_burned'])

array([4.93899949e+08])

In [27]:
np.correlate(fuel_data['fuel_mmbtu_per_unit'], fuel_data['fuel_cost_per_unit_burned'])

array([43347904.22226752])

In [28]:
np.correlate(fuel_data['fuel_cost_per_unit_delivered'], fuel_data['fuel_cost_per_unit_burned'])

array([6.94485051e+10])

In [29]:
np.correlate(fuel_data['fuel_cost_per_mmbtu'], fuel_data['fuel_cost_per_unit_burned'])

array([41853374.17548216])

### Question 9: For the fuel type coal, what is the percentage change in the fuel cost per unit burned in 1998 compared to 1994?

In [43]:
fuel_data.groupby('fuel_type_code_pudl')[['report_year']].count()

report_year
fuel_type_code_pudl             
coal                        8547
gas                        11486
nuclear                      818
oil                         8064
other                        167
waste                        441

In [40]:
fuel_data.groupby('report_year')['fuel_type_code_pudl'].count()

report_year
1994    1235
1995    1201
1996    1088
1997    1094
1998    1107
1999    1050
2000    1373
2001    1356
2002    1205
2003    1211
2004    1192
2005    1269
2006    1243
2007    1264
2008    1228
2009    1222
2010    1261
2011    1240
2012    1243
2013    1199
2014    1171
2015    1093
2016    1034
2017     993
2018     951
Name: fuel_type_code_pudl, dtype: int64

### Question 10: Which year has the highest average fuel cost per unit delivered?

In [54]:
fuel_data.groupby('report_year')['fuel_cost_per_unit_delivered'].max()

report_year
1994      45115.000
1995       2396.770
1996    7793482.000
1997    7964521.000
1998     270604.000
1999       2891.100
2000    1168000.000
2001      34381.000
2002      31416.000
2003      34518.000
2004     132884.000
2005       8651.000
2006       1431.000
2007       4850.872
2008      10490.000
2009     730166.000
2010      44033.320
2011       9545.000
2012      10129.000
2013     135315.000
2014     136410.000
2015     140706.000
2016      69440.000
2017       3584.335
2018     245652.000
Name: fuel_cost_per_unit_delivered, dtype: float64

#### Answer: 1997